serialize suggestion responses as named writeables

andyb-elastic opened the pull request:
This is still in-progress but I wanted to get it up for some feedback. I think the exact location of where some things (e.g. response fields) should go in the class hierarchy still needs to be cleaned up, as well as xcontent generation/parsing. The test coverage also probably could use some work.

The first stage of this will make suggestion responses named writeables and adapt CustomSuggesterIT accordingly, and then the second stage will move that test to an example plugin and make it a rest integration test.

This is related to https://github.com/elastic/elasticsearch/issues/26585

Suggestion responses were previously serialized as streamables which
made writing suggesters in plugins with custom suggestion response types
impossible. This commit makes them serialized as named writeables and
provides a facility for registering a reader for suggestion responses
when registering a suggester.

This also makes Suggestion responses abstract, requiring a suggester
implementation to provide its own types. Suggesters which do not need
anything additional to what is defined in Suggest.Suggestion should
provide a minimal subclass.

elasticmachine commented:
Pinging @elastic/es-search-aggs

jimczi commented on server/src/test/java/org/elasticsearch/search/suggest/CustomSuggesterSearchIT.java:
```diff
@@ -135,7 +149,7 @@ protected XContentBuilder innerToXContent(XContentBuilder builder, Params params
+            return SUGGESTION_NAME;
         }
 
         @Override
```
Is it possible to move this test and the custom suggestion impl in the plugin examples and adds rest tests that checks that xcontent parsing works ? I am not sure that this IT test is enough to validate the plugin integration.
This would be nice to have a real example that shows how to write a custom suggester and the example plugins are the perfect fit for this.

andyb-elastic replied:
I agree on both, that this doesn't completely validate the plugin integration, and that it would be preferable to have it in an example plugin. I was going to make that a follow up PR but I'll add it to this one

andyb-elastic commented on server/src/main/java/org/elasticsearch/plugins/SearchPlugin.java:
```diff
@@ -151,31 +152,59 @@ public ScoreFunctionSpec(String name, Writeable.Reader<T> reader, ScoreFunctionP
+            addSuggestionReader(name, suggestionReader);
+        }
+
+        public SuggesterSpec addSuggestionReader(String writeableName, Writeable.Reader<? extends Suggest.Suggestion> reader) {
```
My thinking here was that it wouldn't be unreasonable for a suggester to return multiple types of suggestions in different situations. For example it looks like some aggregations do this if I'm not mistaken. But  it makes sense and is simpler if they only use one, so I'll do that

andyb-elastic commented:
I added a test for the ability to add custom fields for Entry subclasses. Also changed Suggestion/Entry/Option so that they're all xcontent fragments  and their subclasses can add fields by overriding #toXContent

@jimczi mind taking another look?

jimczi commented on server/src/main/java/org/elasticsearch/search/suggest/Suggest.java:
```diff
@@ -92,6 +90,42 @@ public Suggest(List<Suggestion<? extends Entry<? extends Option>>> suggestions)
 
+    public Suggest(StreamInput in) throws IOException {
+        // in older versions, Suggestion types were serialized as Streamable
+        if (in.getVersion().before(Version.V_6_4_0)) {
```
you need to start with 7.0 and change it when this is backported to 6.5 ?

jimczi commented on server/src/main/java/org/elasticsearch/search/suggest/Suggest.java:
```diff
@@ -92,6 +90,42 @@ public Suggest(List<Suggestion<? extends Entry<? extends Option>>> suggestions)
+                        suggestion = new CompletionSuggestion(in);
+                        break;
+                    case 2: // CompletionSuggestion.TYPE
+                        throw new IllegalArgumentException("Completion suggester 2.x is not supported anymore");
```
I think we can remove this, we don't support 2.x here (nor in 6.x).

jimczi commented on server/src/main/java/org/elasticsearch/search/suggest/Suggest.java:
```diff
@@ -125,42 +159,20 @@ public boolean hasScoreDocs() {
-            out.writeVInt(command.getWriteableType());
-            command.writeTo(out);
+        // in older versions, Suggestion types were serialized as Streamable
+        if (out.getVersion().before(Version.V_6_4_0)) {
```
Same here, version should be 7 for now

andyb-elastic commented:
It looks like there's a bug with serialization across mixed version clusters in the current state of this PR, I'm working on it

andyb-elastic commented:
Jenkins run gradle build tests please

andyb-elastic commented:
Jenkins run gradle build tests please

andyb-elastic commented:
Jenkins run gradle build tests please

andyb-elastic commented:
@jimczi I pushed the serialization tests we talked about

andyb-elastic commented:
Jenkins run gradle build tests please

andyb-elastic commented on server/src/test/java/org/elasticsearch/search/suggest/SuggestTests.java:
```diff
@@ -214,4 +223,39 @@ public void testMergingSuggestionOptions() {
+
+    public void testSerialization() throws IOException {
+        final Version bwcVersion = VersionUtils.randomVersionBetween(random(),
+            Version.CURRENT.minimumCompatibilityVersion(), Version.CURRENT);
```
My thinking was that writing it like this, it will still make sense regardless of what the current version is. For example, there are plenty of tests like this hanging around that explicitly test versions that are now outside of the compatibility range, and I was thinking this is a way of avoiding that

andyb-elastic commented:
Jim and I decided that we don't want to backport this to 6.x because the migration path for plugin authors is a little more involved than I'd originally thought - specifically that authors for who plugin suggesters are currently actually usable are likely using the plain Suggest.Suggestion response type, which this PR makes abstract
